File: examples/spring-boot-maven-plugin/src/main/java/community/flock/wirespec/examples/app/user/LiveUserClient.java
package community.flock.wirespec.examples.app.user;

import community.flock.wirespec.examples.app.common.WirespecHandler;
import community.flock.wirespec.examples.app.common.WirespecSerializer;
import community.flock.wirespec.generated.java.DeleteUserByNameEndpoint;
import community.flock.wirespec.generated.java.GetUserByNameEndpoint;
import community.flock.wirespec.generated.java.GetUsersEndpoint;
import community.flock.wirespec.generated.java.PostUserEndpoint;
import community.flock.wirespec.java.Wirespec;
import org.springframework.stereotype.Component;

import java.util.concurrent.CompletableFuture;

@Component
public class LiveUserClient implements UserClient {

    private final WirespecHandler handler;
    private final Wirespec.ClientEdge<GetUsersEndpoint.Request, GetUsersEndpoint.Response<?>> getUsers;
    private final Wirespec.ClientEdge<GetUserByNameEndpoint.Request, GetUserByNameEndpoint.Response<?>> getUserByName;
    private final Wirespec.ClientEdge<PostUserEndpoint.Request, PostUserEndpoint.Response<?>> postUser;
    private final Wirespec.ClientEdge<DeleteUserByNameEndpoint.Request, DeleteUserByNameEndpoint.Response<?>> deleteUserByName;

    public LiveUserClient(WirespecHandler handler, WirespecSerializer serializer) {
        this.handler = handler;
        this.getUsers = new GetUsersEndpoint.Handler.Handlers().getClient(serializer);
        this.getUserByName = new GetUserByNameEndpoint.Handler.Handlers().getClient(serializer);
        this.postUser = new PostUserEndpoint.Handler.Handlers().getClient(serializer);
        this.deleteUserByName = new DeleteUserByNameEndpoint.Handler.Handlers().getClient(serializer);
    }

    @Override
    public CompletableFuture<GetUsersEndpoint.Response<?>> getUsers(GetUsersEndpoint.Request request) {
        return handler.handle(getUsers.to(request))
                .thenApplyAsync(getUsers::from);
    }

    @Override
    public CompletableFuture<GetUserByNameEndpoint.Response<?>> getUserByName(GetUserByNameEndpoint.Request request) {
        return handler.handle(getUserByName.to(request))
                .thenApplyAsync(getUserByName::from);
    }

    @Override
    public CompletableFuture<PostUserEndpoint.Response<?>> postUser(PostUserEndpoint.Request request) {
        return handler.handle(postUser.to(request))
                .thenApplyAsync(postUser::from);
    }

    @Override
    public CompletableFuture<DeleteUserByNameEndpoint.Response<?>> deleteUserByName(DeleteUserByNameEndpoint.Request request) {
        return handler.handle(deleteUserByName.to(request))
                .thenApplyAsync(deleteUserByName::from);
    }
}

File: examples/wirespec-gradle-plugin-ktor/src/main/kotlin/community/flock/wirespec/examples/app/user/LiveUserClient.kt
package community.flock.wirespec.examples.app.user

import community.flock.wirespec.Wirespec
import community.flock.wirespec.examples.app.common.Serialization
import community.flock.wirespec.generated.kotlin.DeleteUserByNameEndpoint
import community.flock.wirespec.generated.kotlin.GetUserByNameEndpoint
import community.flock.wirespec.generated.kotlin.GetUsersEndpoint
import community.flock.wirespec.generated.kotlin.PostUserEndpoint
import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.request.request
import io.ktor.http.HttpMethod
import io.ktor.http.headers
import io.ktor.http.path
import kotlinx.coroutines.runBlocking

interface UserClient :
    GetUsersEndpoint.Handler,
    GetUserByNameEndpoint.Handler,
    PostUserEndpoint.Handler,
    DeleteUserByNameEndpoint.Handler

class LiveUserClient(private val httpClient: HttpClient = HttpClient()) : UserClient {

    override suspend fun getUsers(request: GetUsersEndpoint.Request) =
        with(GetUsersEndpoint.Handler.client(Serialization)) {
            externalize(request).let(::handle).let(::internalize)
        }

    override suspend fun getUserByName(request: GetUserByNameEndpoint.Request) =
        with(GetUserByNameEndpoint.Handler.client(Serialization)) {
            externalize(request).let(::handle).let(::internalize)
        }

    override suspend fun postUser(request: PostUserEndpoint.Request) =
        with(PostUserEndpoint.Handler.client(Serialization)) {
            externalize(request).let(::handle).let(::internalize)
        }

    override suspend fun deleteUserByName(request: DeleteUserByNameEndpoint.Request) =
        with(DeleteUserByNameEndpoint.Handler.client(Serialization)) {
            externalize(request).let(::handle).let(::internalize)
        }

    private fun handle(request: Wirespec.RawRequest): Wirespec.RawResponse = runBlocking {
        val response = httpClient.request("http://localhost:8080/") {
            method = HttpMethod.parse(request.method)
            url {
                path(*request.path.toTypedArray())
            }
            headers {
                request.headers.forEach { (t, u) -> appendAll(t, u) }
            }
        }
        response.run {
            Wirespec.RawResponse(
                statusCode = status.value,
                headers = headers.entries().associate { it.key to it.value },
                body = body()
            )
        }
    }
}
